src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs

```diff
@@ -39,5 +40,45 @@ public static Span<T> AsSpan<T>(List<T>? list)
         /// <remarks>Items should not be added to or removed from the <see cref="Dictionary{TKey, TValue}"/> while the ref <typeparamref name="TValue"/> is in use.</remarks>
         public static ref TValue? GetValueRefOrAddDefault<TKey, TValue>(Dictionary<TKey, TValue> dictionary, TKey key, out bool exists) where TKey : notnull
             => ref Dictionary<TKey, TValue>.CollectionsMarshalHelper.GetValueRefOrAddDefault(dictionary, key, out exists);
+
+        /// <summary>
+        /// Sets the count of the <see cref="List{T}"/> to the specified value.
+        /// </summary>
+        /// <param name="list">The list to set the count of.</param>
+        /// <param name="count">The value to set the list's count to.</param>
+        /// <exception cref="NullReferenceException">
+        /// <paramref name="list"/> is <see langword="null"/>.
+        /// </exception>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// <paramref name="count"/> is negative.
+        /// </exception>
+        /// <remarks>
+        /// When increasing the count, uninitialized data is being exposed.
+        /// </remarks>
+        public static void SetCount<T>(List<T> list, int count)
+        {
+            if (count < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException_NeedNonNegNum(nameof(count));
+            }
+
+            if (count == list._size)
+            {
+                return;
+            }
+
+            list._version++;
+
+            if (count > list.Capacity)
+            {
+                list.Grow(count);
+            }
+            else if (count < list._size && RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+            {
+                Array.Clear(list._items, count, list._size - count);
```

Comment:
If this is in `CollectionsMarshal` vs say having a setter on `.Count` does the array want to be cleared?

e.g. it could enable
```csharp
var preCount = list.Count;
// Only include first item for api that takes `List<T>`
CollectionsMarshal.SetCount(list, 1);
MyListApi((IReadOnlyList)list);
//Restore all previous elements
CollectionsMarshal.SetCount(list, preCount);
```

Reply:
That was discussed (and unresolved) in https://github.com/dotnet/runtime/pull/77794#discussion_r1012105187.

Reply:
Ok, but the way it currently behaves would be covered by adding a `set` to `.Count`; e.g. its a very safe method and replicates the settable `.length` on JS arrays for example

Reply:
I don't have a particular preference. This is an unsafe API and users can easily do the wrong thing, so if there are places where not clearing are beneficial then it may be worth exposing.

We'd just want to be extra explicit in the docs that not clearing can lead to "leaks" in that the GC will think there is still a live reference.

Reply:
> Ok, but the way it currently behaves would be covered by adding a set to .Count; e.g. its a very safe method

We wouldn't consider arbitrary data show up in the list by increasing the count to be "safe", and that could happen if you, for example, add some data, then remove some data, and then increase the count... the data that was previously there but explicitly removed might be back again.

Reply:
Unsafe apis are to allow users to potentially do the wrong thing

I'd personally skip the clearing in `CollectionsMarshal` then add setter to `.Count` that has clearing
```csharp
public int Count 
{
    get => _size;
    set 
    {
         // Resize Count but with clearing
    }
}
```
Then the `CollectionsMarshal` parallel has the unsafe component (if you need to temporally reduce size, then increase after but not loose elements)

Reply:
> > Ok, but the way it currently behaves would be covered by adding a set to .Count; e.g. its a very safe method
> 
> We wouldn't consider arbitrary data showing up in the list by increasing the count to be "safe", ...

So its an odd mix currently 🤷‍♂️ Add safe api on List going all the way on clearing; then go all the way here on not clearing?

Reply:
> then add setter to .Count that has clearing

It'd need to clear even when increasing the count, regardless of whether T is or contains reference types.

If we want to do that, I have no problem with it, but it should be a separate API proposal.

> Add safe api on List going all the way on clearing; then go all the way here on not clearing?

I don't want this discussion to derail this API again.
https://github.com/dotnet/runtime/pull/77794#discussion_r1089233292
"In the name of making forward progress here, let's finish this PR with the behavior of supporting shrinking and clearing, and then we can have a follow-up conversation about whether to change the behavior prior to shipping."

Reply:
> If this is in `CollectionsMarshal` vs say having a setter on `.Count` does the array want to be cleared?
> 
> e.g. it could enable
> 
> ```cs
> var preCount = list.Count;
> // Only include first item for api that takes `List<T>`
> CollectionsMarshal.SetCount(list, 1);
> MyListApi((IReadOnlyList)list);
> //Restore all previous elements
> CollectionsMarshal.SetCount(list, preCount);
> ```

Seeing that any API like `Add` could cause the data to be overwritten or lost due to a reallocation, I'd only consider not zeroing it for performance reasons.

And for now I think that accidental memory leaks would be more troubling here so I'm going to leave it as is, as it has been agree on.